modules/rgbd/src/hash_tsdf.hpp

```diff
@@ -60,53 +137,147 @@ struct tsdf_hash
+        return volumeUnit->at(volUnitLocalIdx);
+    }
 
     //! Return the voxel given the point in volume coordinate system i.e., (metric scale 1 unit =
     //! 1m)
-    virtual TsdfVoxel at(const cv::Point3f& point) const;
+    TsdfVoxel at_(const Point3f& point) const
+    {
+        Vec3i volumeUnitIdx              = volumeToVolumeUnitIdx(point);
+        VolumeUnitMap::const_iterator it = volumeUnits.find(volumeUnitIdx);
+        if (it == volumeUnits.end())
+        {
+            TsdfVoxel dummy;
+            dummy.tsdf   = 1.f;
+            dummy.weight = 0;
+            return dummy;
+        }
+        Ptr<TSDFVolumeCPU> volumeUnit = std::dynamic_pointer_cast<TSDFVolumeCPU>(it->second.pVolume);
+
+        Point3f volumeUnitPos = volumeUnitIdxToVolume(volumeUnitIdx);
+        Vec3i volUnitLocalIdx = volumeToVoxelCoord(point - volumeUnitPos);
+        volUnitLocalIdx       = Vec3i(abs(volUnitLocalIdx[0]), abs(volUnitLocalIdx[1]), abs(volUnitLocalIdx[2]));
+        return volumeUnit->at(volUnitLocalIdx);
+    }
 
-    inline TsdfType interpolateVoxel(const cv::Point3f& point) const;
-    Point3f getNormalVoxel(cv::Point3f p) const;
+    TsdfType interpolateVoxel_(const Point3f& point) const
+    {
+        Point3f neighbourCoords[] = { Point3f(0, 0, 0), Point3f(0, 0, 1), Point3f(0, 1, 0), Point3f(0, 1, 1),
+                                      Point3f(1, 0, 0), Point3f(1, 0, 1), Point3f(1, 1, 0), Point3f(1, 1, 1) };
+
+        int ix = cvFloor(point.x);
+        int iy = cvFloor(point.y);
+        int iz = cvFloor(point.z);
+
+        float tx = point.x - ix;
+        float ty = point.y - iy;
+        float tz = point.z - iz;
+
+        TsdfType vx[8];
+        for (int i = 0; i < 8; i++)
+            vx[i] = at(neighbourCoords[i] * voxelSize + point).tsdf;
+
+        TsdfType v00 = vx[0] + tz * (vx[1] - vx[0]);
+        TsdfType v01 = vx[2] + tz * (vx[3] - vx[2]);
+        TsdfType v10 = vx[4] + tz * (vx[5] - vx[4]);
+        TsdfType v11 = vx[6] + tz * (vx[7] - vx[6]);
+
+        TsdfType v0 = v00 + ty * (v01 - v00);
+        TsdfType v1 = v10 + ty * (v11 - v10);
+
+        return v0 + tx * (v1 - v0);
+    }
+
+    Point3f getNormalVoxel_(const Point3f& point) const
+    {
+        Vec3f pointVec(point);
+        Vec3f normal = Vec3f(0, 0, 0);
+
+        Vec3f pointPrev = point;
+        Vec3f pointNext = point;
+
+        for (int c = 0; c < 3; c++)
+        {
+            pointPrev[c] -= voxelSize;
+            pointNext[c] += voxelSize;
+
+            normal[c] = at(Point3f(pointNext)).tsdf - at(Point3f(pointPrev)).tsdf;
+            /* normal[c] = interpolateVoxel(Point3f(pointNext)) - interpolateVoxel(Point3f(pointPrev)); */
+
+            pointPrev[c] = pointVec[c];
+            pointNext[c] = pointVec[c];
+        }
+        // std::cout << normal << std::endl;
+        float nv = sqrt(normal[0] * normal[0] + normal[1] * normal[1] + normal[2] * normal[2]);
+        return nv < 0.0001f ? nan3 : normal / nv;
+    }
```

Comment:
While `interpolateVoxel_` and `getNormalVoxel_` are of their current size, it's OK to keep them in the header file. But they are going to be rewritten for performance purposes, the code will be more complex, so it'd be better to move them into `.cpp` file.

Reply:
Yes, you have a point. In our earlier conversation on slack, we noticed that there was a compilation issue with inline functions and definition in cpp file. So I had moved it here. Since we expect these functions to get larger, I have moved them back into cpp file.